### DISCUSSION TITLE
Too many open files

### DISCUSSION THREAD

**Discussion by KIONIASRCOE** [eclipse-milo/milo#1352] [General]
**Describe the bug**
Unstable network or node connection error cause milo crashed.

**Expected behavior**
I cannot solve the issue when a network broke several seconds. 

**Logs and Packet Captures**
![milo](https://github.com/user-attachments/assets/b216329a-57f5-4653-bb87-bcb69a5770da)

**Additional context**
OPC UA server is Siemens S7-1500

I wonder what is the security tmp file means and for what purpose?

**Comment by kevinherron** [comment]
KeyStoreLoader is your code. I assume it's based on [this](https://github.com/eclipse-milo/milo/blob/master/milo-examples/client-examples/src/main/java/org/eclipse/milo/examples/client/KeyStoreLoader.java) from the examples, but this copy doesn't log.

The other longer stack you're highlighting doesn't have any Milo code in it. Look at your own code to see what's happening. (`DeviceService292Impl.java:61`?)

**Reply by KIONIASRCOE** [reply]
Hi Kevin,

Every time when my client code reconnect OPC UA server, it will open the milo-client.pfx file again and the source code is here. I want to have a connection pool could be reused, but the milo-client.pfx is opened too many times, so the system crashed. I wonder is there any suggestion how to have a pool and I don't need client to trust the server again and again.

public PooledObject<OpcUaClient> makeObject(MiloProperties.Config key) throws Exception {
    OpcUaClient client = null;
    try {
        client = createClient(key);
        client.connect().get();
        return new DefaultPooledObject<>(client);
    } catch (Exception e) {
        log.error("makeObject Exception"+e.getMessage(), e);
        if (client != null) {
            client.disconnect().get();
        }
        throw new InterruptedException(e.getMessage());
    }
}

**Reply by kevinherron** [reply]
You don't have to open it every time you create a client; you can open it once and use the same instance each time you create a client. 

You should also verify this is even the cause of your file leak. If you copied the KeyStoreLoader from the example entirely the InputStream should be getting closed: https://github.com/eclipse-milo/milo/blob/c6a875c2fa8f65c48815410d6bbf02d43560bcd4/milo-examples/client-examples/src/main/java/org/eclipse/milo/examples/client/KeyStoreLoader.java#L85-L87

**Reply by KIONIASRCOE** [reply]
Thanks, I'll try.
